kale/embed/mpca.py

```diff
@@ -72,200 +72,220 @@ class MPCA(BaseEstimator, TransformerMixin):
-        n_spl = shape_[-1]  # number of samples
-        self.ndim = X.ndim
-        self.shape_in = shape_[:-1]
-        self.mean_ = np.mean(X, axis=-1)
-        X_ = np.zeros(X.shape)
+        shape_ = x.shape  # shape of input data
+        n_spl = shape_[0]  # number of samples
+        n_dim = x.ndim
+
+        self.shape_in = shape_[1:]
+        self.mean_ = np.mean(x, axis=0)
+        x = x - self.mean_
+
         # init
-        Phi = dict()
-        eig_vecs_sorted = dict()  # dictionary of eigenvectors for all modes/orders
-        lambdas = dict()  # dictionary of eigenvalues for all modes/orders
+        phi = dict()
+        # ev_sorted = dict()  # dictionary of eigenvectors for all modes/orders
+        # lambdas = dict()  # dictionary of eigenvalues for all modes/orders
         # dictionary of cumulative distribution of eigenvalues for all modes/orders
-        cums = dict()
-        proj_mats = []
+        # cums = dict()
         shape_out = ()
-        for i in range(n_spl):
-            X_[..., i] = X[..., i] - self.mean_
-            for j in range(self.ndim - 1):
-                X_i = unfold(X_[..., i], mode=j)
-                if j not in Phi:
-                    Phi[j] = 0
-                Phi[j] = Phi[j] + np.dot(X_i, X_i.T)
-
-        for i in range(self.ndim - 1):
-            eig_vals, eig_vecs = np.linalg.eig(Phi[i])
-            idx_sorted = eig_vals.argsort()[::-1]
-            eig_vecs_sorted[i] = eig_vecs[:, idx_sorted]
-            cum = eig_vals[idx_sorted]
+        proj_mats = []
+
+        for i in range(1, n_dim):
+            for j in range(n_spl):
+                # unfold the j_th tensor along the i_th mode
+                x_ji = unfold(x[j, :], mode=i - 1)
+                if i not in phi:
+                    phi[i] = 0
+                phi[i] = phi[i] + np.dot(x_ji, x_ji.T)
+
+        for i in range(1, n_dim):
+            eig_values, eig_vectors = np.linalg.eig(phi[i])
+            idx_sorted = (-1 * eig_values).argsort()
+            # ev_sorted[i] = eig_vectors[:, idx_sorted]
+            cum = eig_values[idx_sorted]
             var_tot = np.sum(cum)
 
-            for j in range(cum.shape[0]):
+            for j in range(1, cum.shape[0] + 1):
                 if np.sum(cum[:j]) / var_tot > self.var_ratio:
-                    shape_out += (j + 1,)
+                    shape_out += (j,)
                     break
-            cums[i] = cum
+            # cums[i] = cum
+            proj_mats.append(eig_vectors[:, idx_sorted][:, : shape_out[i - 1]].T)
 
-            proj_mats.append(eig_vecs_sorted[i][:, : shape_out[i]].T)
+        if self.return_vector and isinstance(self.n_components, int):
+            if self.n_components > np.prod(shape_out):
+                self.n_components = np.prod(shape_out)
+                warn_msg = "n_components exceeds the maximum number of components, " \
+                           "all components will be returned instead."
```

Comment:
I think the message can be more compact <120) to that line break won't be needed.